### DISCUSSION TITLE
Adding more filtering operations, possibly behind feature

### DISCUSSION THREAD

**Discussion by ginger51011** [obi1kenobi/trustfall#276] [Ideas]
Hello!

I have come across a type of filtering I would like to implement, but I am not sure if we want to "pollute" the operations, since you technically can use edge parameters for more advanced filtering.

My idea is to allow semantic versioning operations on strings, but perhaps this is too much of an edge case. Perhaps behind a feature?

What do you think @obi1kenobi ?

**Comment by obi1kenobi** [comment]
What timescale do you need this on?

The "real" long-term solution here is custom scalar types. Imagine that in your schema you could define a new scalar like `scalar Semver`, and then can define vertex properties like `version: Semver`. The scalar type definition would specify which kinds of operators are allowed, and the adapter implementation would implement those operators. This would allow implementing version-like comparisons like `1.5.0 < 1.23.4`.

Of course this long-term solution requires a decent bit of work. I have a design in mind, but it still needs to be implemented and tested throughout. If you just need a one-off quickly, forking the project and adding your own filter operators is probably the fastest option.

Adding semver-specific operators behind a feature sounds problematic from an ergonomics and usability perspective: e.g. nothing prevents people from using semver operators on non-semver string fields, so what happens when someone does that? Even if the feature is unstable and off by default, it still carries a maintenance cost. If your employer is willing to financially support Trustfall to compensate for the ongoing maintenance burden of the feature, this is something I'd happily consider. Without such an arrangement, that maintenance burden would be too great and would take time and energy I'd rather dedicate to other things.

**Comment by ginger51011** [comment]
Thanks for the answer!

I see your point, and I suspected that you had a (better) long term solution in mind, so I felt like I had to ask.

For me it was mostly a one-off thing, an idea I was experimenting with, so the timescale is basically _Maybe I can do a quick solution in a couple of hours_ but yes I might do it on my fork but probably not.

Thanks for taking the time! Trustfall has proven to be very useful, and hopefully I can release what I'm working on to the public soon 😊

**Reply by obi1kenobi** [reply]
The quickest solution, depending on the exact details you need, would be one of:
- hack it into the schema as some sort of edge parameter, or
- just output the property value and apply semver-aware operators afterward

The next quickest would be forking and adding the custom operator, or just editing the comparison operator implementation for strings to do semver-aware comparisons instead, if you never compare non-semver strings with those operators.

Hope that's useful!

I'm really glad to hear Trustfall has been helpful, and I look forward to seeing what you've been working on! 🙌
